### DISCUSSION TITLE
How do you use I2S.onReceive(function) ?

### DISCUSSION THREAD

**Discussion by frohro** [earlephilhower/arduino-pico#2097] [Q&A]
Hi All,
I am trying to set up ping pong buffers to read I2S input and send it over UDP using interrupts or callbacks.  It looks like the correct function to use is I2S.onReceive().  I am unsure if I am supposed to read buffers already set up in the I2S software or if I am supposed to use my own.  I have guessed some code using my own, but I don't .  See below.  I see some packets with wireshark, but only a few (30 kBytes/s, significantly less than I got [here](https://github.com/earlephilhower/arduino-pico/discussions/2093),  not the number I expect.  The flag dataReady does get set.  I am also confused about how I'm supposed to use it without using the blocking read() call that it warns not to use in the [Arduino-Pico Docs](https://arduino-pico.readthedocs.io/en/latest/i2s.html#void-onreceive-void-fn-void).  It seems it should be okay because I already know there is data waiting for me, so it should not block, but I'm probably wrong.  Can anyone help?

```
/*
 * This Pico W program is a stab at connecting to the PCM1808 via I2S and reading the data at RATE samples per second.
 * It then sends the data over WIFI using UDP using ping pong buffers. 
 *
 * Author: Rob Frohne, KL7NA, with help from Github Copilot.
 * 4/9/2024
 */
#include <Arduino.h>
#include <I2S.h>
#include <WiFi.h>
#include <WiFiUdp.h>
#include <si5351.h>

#ifndef STASSID
#define STASSID Frohne - 2.4GHz
#endif

#define RATE 96000
#define MCLK_MULT 256 // 384 for 48 BCK per frame,  256 for 64 BCK per frame
#define BUFFER_SIZE 1400

I2S i2s(INPUT);

WiFiUDP udp;
const char *udpAddress = "192.168.1.101"; // Put your laptop IP here.
const unsigned int udpPort = 12345;

char bufferA[BUFFER_SIZE];
char bufferB[BUFFER_SIZE];
char *currentBuffer = bufferA;
char *sendBuffer = bufferB;
int bufferIndex = 0;
volatile bool dataReady = false;

void i2sDataReceived()
{
    static int32_t r, l;
    i2s.read32(&l, &r); // Read the next l and r values
    l = l << 9;
    r = r << 9;
    // Fill the current buffer with data
    int n = snprintf(currentBuffer + bufferIndex, BUFFER_SIZE - bufferIndex, "%d,%d\r\n", l, r);
    bufferIndex += n;

    // If the buffer is full, swap the buffers
    if (bufferIndex >= BUFFER_SIZE)
    {
        char *temp = currentBuffer;
        currentBuffer = sendBuffer;
        sendBuffer = temp;
        bufferIndex = 0;

        // Set the flag to indicate data is ready to be sent
        dataReady = true;
\_t *)sendBuffer, BUFFER_SIZE);
        udp.endPacket();
    }
}

void setup()
{
    Serial.begin();

    i2s.setDATA(2); // These are the pins for the data on the SDR-TRX
    i2s.setBCLK(0);
    i2s.setMCLK(3);
    // Note: LRCK pin is BCK pin plus 1 (1 in this case).
    i2s.setBitsPerSample(24);
    i2s.setFrequency(RATE);
    i2s.setMCLKmult(MCLK_MULT);
    i2s.setSysClk(RATE);
    i2s.setBuffers(32, 0, 0);

    WiFi.begin("Frohne-2.4GHz");

    while (WiFi.status() != WL_CONNECTED)
    {
        delay(10000);
        Serial.println("Trying to connect to WiFi...");
    }
    Serial.println("WiFi connected");
    Serial.print("IP address: ");
    Serial.println(WiFi.localIP());

    udp.begin(12345);

    // Set the callback function
    i2s.onReceive(i2sDataReceived);
    i2s.begin();
}

void loop()
{
    // If data is ready, send it over UDP
    if (dataReady)
    {
        udp.beginPacket(udpAddress, udpPort);
        udp.write((const uint8_t *)sendBuffer, BUFFER_SIZE);
        udp.endPacket();
        dataReady = false;
    }

    // Your other main loop code goes here...
}
```

**Comment by earlephilhower** [comment]
I'm traveling this week so can't really get into your code on my phone but I'll give you a few tips.


Yes, you need to use your own buffers.  The I2S ones are owned by the DMA engines and can't be accessed by you.

The read call will only block of there are no samples available.  From a callback you are guaranteed at least one buffer is ready so you can blindly read that many samples.

There is also a non blocking version of read, check the I2S header.

Your setBuffers call looks wrong, please verify you're seeing the right number of samples and buffers.  One of those looks to be set to 0 which is going to hurt when the defaults are used instead.
